include/bout/region.hxx

```diff
@@ -0,0 +1,236 @@
+  ind2D& operator=(const int& i){ind=i;return *this;}
+};
+
+template<typename T = ind3D>
+class Region {
+  //Following prevents a Region being created with anything other
+  //than ind2D or ind3D as template type
+  static_assert(std::is_base_of<ind2D,T>::value || std::is_base_of<ind3D,T>::value, "Region must be templated with either ind2D or ind3D");
+  
+public:
+  typedef T data_type;
+
+  /// Indices to iterate over
+  typedef std::vector<T> regionIndices;
+  /// Start and end of contiguous region
+  typedef std::pair<T, T> contiguousBlock;
+  /// Collection of contiguous regions
+  typedef std::vector<contiguousBlock> contiguousBlocks;
+
+  //NOTE::
+  // Probably want to require a mesh in constructor, both to know nx/ny/nz o
+  // but also to ensure consistency etc.
+  
+  //Want to construct from:
+  // * vector of ints
+  // * existing region
+  // * start/stop of each dimension.
+
+  //Want to make this private to disable but think it may be needed as we put Regions into
+  //maps which seems to need to be able to make "empty" objects.
+  Region<T>(){};
+
+  Region<T>(int xstart, int xend, int ystart, int yend,
+	    int zstart, int zend, int ny, int nz){
+    indices = createRegionIndices(xstart, xend, ystart, yend, zstart, zend, ny, nz);
+    blocks  = getContiguousBlocks();
+  };
+  
+  Region<T>(regionIndices& indices): indices(indices){
+    blocks = getContiguousBlocks();
+  };
+
+  /// Destructor
+  ~Region(){};
+
+  contiguousBlocks getBlocks(){return blocks;};//Setter not appropriate
+  regionIndices getIndices(){return indices;};//Setter could be ok
+
+  // TODO: Should be able to add regions (would just require extending
+  // indices and recalculating blocks). This raises question of should
+  // we be able to subtract regions, and if so what does that mean.
+  // Addition could be simple and just extend or we could seek to
+  // remove duplicate points. Former probably mostly ok. Note we do
+  // want to allow duplicate points (one reason we use vector and
+  // not set) but what if we add a region that has some duplicates?
+  // We could retain them but common usage would probably not want
+  // the duplicates.
+
+  // We could sort indices (either forcibly or on request) to try
+  // to ensure most contiguous+ordered access. Probably ok in common
+  // use but would cause problems if order is important, for example
+  // if we could use regions to transpose/reorder data. If regions were
+  // sorted this would prevent this usage.
+  
+private:
+  regionIndices indices;  //Flattened indices
+  contiguousBlocks blocks;//Contiguous sections of flattened indices
+
+  /// Helper function to create a regionIndices, given the start and end
+  /// points in x, y, z, and the total y, z lengths
+  inline regionIndices createRegionIndices(int xstart, int xend, int ystart, int yend,
+  					   int zstart, int zend, int ny, int nz){
+    int len = (xend - xstart + 1) * (yend - ystart + 1) * (zend - zstart + 1);
```

Comment:
no check that `xend > xstart +1` etc.

Reply:
Yes, those would be sensible tests. I'll add them.